src/blueapi/cli/cli.py

```diff
@@ -32,24 +33,17 @@
     "-c", "--config", type=Path, help="Path to configuration YAML file", multiple=True
 )
 @click.pass_context
-def main(ctx: click.Context, config: Path | None | tuple[Path, ...]) -> None:
-    # if no command is supplied, run with the options passed
-
-    config_loader = ConfigLoader(ApplicationConfig)
-    if config is not None:
-        configs = (config,) if isinstance(config, Path) else config
-        for path in configs:
-            if path.exists():
-                config_loader.use_values_from_yaml(path)
-            else:
-                raise FileNotFoundError(f"Cannot find file: {path}")
-
-    ctx.ensure_object(dict)
-    loaded_config: ApplicationConfig = config_loader.load()
-
-    ctx.obj["config"] = loaded_config
+def main(ctx: click.Context, config: PathType) -> None:
+    # Override default yaml_file path in the model_config if `config` is provided
+    ApplicationConfig.model_config["yaml_file"] = config
+    app_config = ApplicationConfig()  # Instantiates with customized sources
```

Comment:
Should: I'm concerned this. mutation of global state is an antipattern. Could you instead subclass the config?

```suggestion
    class YamlConfig(ApplicationConfig):
        model_config = {**ApplicationConfig.model_config, **{"yaml_file": config}}
    app_config = YamlConfig()
```

Reply:
that's how Pydantic uses it and this is not regular state but settings

also this is not Java

Reply:
Mutation of global state is an antipattern in any programming language ([good thread](https://softwareengineering.stackexchange.com/questions/148108/why-is-global-state-so-evil)).

Now I have to remember to address this if I use `ApplicationConfig` somewhere else, just like you have:

```python
ApplicationConfig.model_config["yaml_file"] = None
```

There is nothing that makes this obvious to me as a developer.

Reply:
the mutable settings option is here:
and we're not using it
https://docs.pydantic.dev/latest/concepts/pydantic_settings/#removing-sources

Reply:
It looks like what we're trying to do was just not a well-envisioned use case: https://github.com/pydantic/pydantic-settings/issues/346

Should maybe reconsider the use of `BaseSettings` and/or raise a PR upstream?

Reply:
that is true, turns out that pydantic-settings isn't as well-finished product like pydantic itself